CCheckQueue Unit Tests

JeremyRubin opened the pull request:
This PR builds on https://github.com/bitcoin/bitcoin/pull/9495 to unit test the CCheckQueue for correctness.

The cases covered in these tests are:

1. Standard usage
2. Failing checks are caught
3. Prior blocks failing don't interfere with future blocks
4. No Memory leakage (all check destructors are called before new blocks allowed, memory is freed).
5. Thread Safety

JeremyRubin committed:
Add CheckQueue Tests

JeremyRubin committed:
Fix CCheckQueue IsIdle (potential) race condition and remove dangerous constructors.

instagibbs commented on src/test/checkqueue_tests.cpp:
```diff
@@ -0,0 +1,395 @@
+// Copyright (c) 2012-2015 The Bitcoin Core developers
```
couple of years out of date :)

JeremyRubin replied:
will fix!

JeremyRubin commented:
Sorry for the line noise; the earlier build error should be addressed now.

ryanofsky commented on src/test/checkqueue_tests.cpp:
```diff
@@ -0,0 +1,402 @@
+ */
+void Correct_Queue_range(std::vector<size_t> range)
+{
+    auto small_queue = std::shared_ptr<Correct_Queue>(new Correct_Queue {128});
```
Why is small_queue a shared pointer, not a unique pointer or just plain stack variable? Maybe add a comment explaining. Also, you could probably use make_shared if it does need to be a shared pointer.

ryanofsky commented on src/test/checkqueue_tests.cpp:
```diff
@@ -0,0 +1,402 @@
+            }
+            result[end_fails ? 0 : 1] = control.Wait();
+        }
+        BOOST_CHECK(!result[0]);
```
Would seem more direct to just BOOST_CHECK the control.Wait() call instead of putting the results in an intermediate array. This way is ok too, though.

ryanofsky commented on src/test/checkqueue_tests.cpp:
```diff
@@ -0,0 +1,402 @@
+    auto small_queue = std::shared_ptr<Correct_Queue>(new Correct_Queue {128});
+    boost::thread_group tg;
+    for (auto x = 0; x < nScriptCheckThreads; ++x) {
+       tg.create_thread([=]{small_queue->Thread();});
```
Replacing [=] with [&] might allow small_queue not to be a shared_ptr.

ryanofsky commented on src/test/checkqueue_tests.cpp:
```diff
@@ -0,0 +1,402 @@
+    bool r = true;
+    for (size_t i = 0; i < COUNT; ++i)
+        r = r && UniqueCheck::results.count(i) == 1;
+    BOOST_REQUIRE(r);
```
Maybe also check that UniqueCheck::results.size == COUNT.

ryanofsky commented on src/test/checkqueue_tests.cpp:
```diff
@@ -0,0 +1,402 @@
+
+// Test that blocks which might allocate lots of memory free their memory agressively.
+//
+// This test attempts to catch a pathalogical case where by lazily freeing
```
pathological (spelling)

ryanofsky commented on src/test/checkqueue_tests.cpp:
```diff
@@ -0,0 +1,402 @@
+// checks might mean leaving a check un-swapped out, and decreasing by 1 each
+// time could leave the data hanging across a sequence of blocks.
+//
+// This test (failing) is dependent on not being able to handle
```
Instead of having the test be nondeterministic in this way, would anything be lost if you had the MemoryCheck constructor increment a static counter when passed a true arg, and the MemoryCheck destructor decrement the counter if the object was constructed with a true arg. Then you could detect the error case explicitly by checking the counter, and not have to allocate big chunks of memory.

ryanofsky commented on src/test/checkqueue_tests.cpp:
```diff
@@ -0,0 +1,402 @@
+    //
+    // Note that this cannot cause a spurious failure, only could mean that
+    // the test doesn't actually end up checking that control waited.
+    MilliSleep(1000);
```
I think you could easily make this test deterministic as well, eliminating the long sleep and the while (frozen) busy loops. Would just need to have ~FrozenCleanupCheck increment a counter and signal a conditional variable so you could wait here for enough jobs to be frozen, and then do the boost check. Then this could signal another condition variable to unfreeze the jobs. I think it would be worth changing this to make the test more efficient and reliable.

ryanofsky reviewed (commented):
Overall great and very clever tests (thread safety one was fun to figure out). Added a bunch of minor comments. The only two comments I would really urge you to consider are the ones on the Memory and FrozenCleanup tests. I think it would be good to check same conditions without allocating large chunks of memory or doing 1-second busy loops so these tests can be more efficient and more reliable.

ryanofsky commented on src/test/checkqueue_tests.cpp:
```diff
@@ -372,29 +381,40 @@ BOOST_AUTO_TEST_CASE(test_CheckQueueControl_Locks)
-        while (!done && !fails2) {
-            fails2 = queue->ControlMutex.try_lock();
+        std::mutex m;
+        bool has_lock {false};
```
It might be clearer to replace all these bools with an enum like { STARTED, TRY_LOCK, TRY_LOCK_DONE, DONE, DONE_ACK }.

JeremyRubin replied:
Going to ignore this, I don't think it makes it more clear (to me, it's easier to debug several variables). If someone disagrees strongly, will change.

ryanofsky reviewed (commented):
utACK d44af13ddc3615686a3e76cf9a3412999db0d692. Left one minor comment, feel free to ignore.

laanwj commented:
ACK, needs squashing

JeremyRubin commented:
Squashed!

JeremyRubin commented:
Rebased to be on top of https://github.com/bitcoin/bitcoin/pull/9495.

kallewoof commented on src/test/checkqueue_tests.cpp:
```diff
@@ -0,0 +1,439 @@
+                control.Add(vChecks);
+            }
+        }
+        BOOST_REQUIRE(MemoryCheck::fake_allocated_memory == 0);
```
The MemoryCheck struct destructor does not --, so this should not be == 0.

JeremyRubin replied:
Yes, it was the latter. The for loop never made anything (i = 9999; i<9999). Will fix :)

kallewoof commented on src/test/checkqueue_tests.cpp:
```diff
@@ -0,0 +1,439 @@
+    };
+    ~MemoryCheck(){
+        if (b) {
+            fake_allocated_memory += 1;
```
`fake_allocated_memory -= 1`

kallewoof commented on src/test/checkqueue_tests.cpp:
```diff
@@ -0,0 +1,439 @@
+BOOST_AUTO_TEST_CASE(test_CheckQueue_Recovers_From_Failure)
+{
+    auto fail_queue = std::unique_ptr<Failing_Queue>(new Failing_Queue {QUEUE_BATCH_SIZE});
+    std::array<FailingCheck, 100> checks;
```
Unused variable.

kallewoof commented on src/test/checkqueue_tests.cpp:
```diff
@@ -0,0 +1,439 @@
+        // Wait until the queue has finished all jobs and frozen
+        FrozenCleanupCheck::cv.wait(l, [](){return FrozenCleanupCheck::nFrozen == 1;});
+        // Try to get control of the queue a bunch of times
+        for (auto x = 0; x < 100; ++x) {
```
Doing `!fails && x < 100` here and simply `fails = queue->ControlMutex.try_lock();` would break iteration on first fail rather than iterate over all 100 (e.g. if first `try_lock()` fails).

JeremyRubin replied:
sure..

kallewoof commented on src/test/checkqueue_tests.cpp:
```diff
@@ -0,0 +1,439 @@
+            // Wait for thread to get the lock
+            cv.wait(l, [&](){return has_lock;});
+            bool fails = false;
+            for (auto x = 0; x < 100; ++x) {
```
Same here with `!fails && x < 100` as above.

JeremyRubin commented:
Fixed the issues that @kallewoof raised, and squashed.

Unsquashed preserved here https://github.com/JeremyRubin/bitcoin/tree/checkqueue-tests-unsquashed.

kallewoof commented:
utACK 96c7f2c

I'm a bit concerned about non-deterministic behavior in tests as this tends to be a pain when you do run into a problem. Or is this fixed seed / PRNG so that the numbers are always the same each time?

JeremyRubin commented:
I could make them deterministic if that's desirable, but realistically these tests are already non-deterministic by virtue of being multithreaded. None of the uses of GetRand are particularly dangerous here, although perhaps they area a little slower than could be.

kallewoof commented:
I think that would be desirable, even if the multithreading makes it not 100%.

kallewoof commented:
ACK 96c7f2c
